### ISSUE TITLE
Network Issues.

### ISSUE THREAD

**Issue by Ferouden** [Microsoft/WSL#2724]
On 1709, subsystem `ubuntu` and `archlinux`.
When I use ping .
`$ /usr/bin/time -v ping www.bing.com -c 1`
```
PING cn-0001.cn-msedge.net (202.89.233.101) 56(84) bytes of data.
64 bytes from 202.89.233.101 (202.89.233.101): icmp_seq=1 ttl=108 time=42.8 ms

--- cn-0001.cn-msedge.net ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 42.869/42.869/42.869/0.000 ms
0.00user 0.00system 0:10.10elapsed 0%CPU (0avgtext+0avgdata 1824maxresident)k
0inputs+0outputs (0major+571minor)pagefaults 0swaps
```

elapsed time is **10** sec! 
And I noticed that I got it's ip very slow.

And then I tried nslookup.
`$ /usr/bin/time nslookup www.bing.com`
```
Server:         115.159.146.99
Address:        115.159.146.99#53

Non-authoritative answer:
www.bing.com    canonical name = cn.cn-0001.cn-msedge.net.
cn.cn-0001.cn-msedge.net        canonical name = cn-0001.cn-msedge.net.
Name:   cn-0001.cn-msedge.net
Address: 202.89.233.101
Name:   cn-0001.cn-msedge.net
Address: 202.89.233.100

0.01user 0.00system 0:00.14elapsed 10%CPU (0avgtext+0avgdata 5960maxresident)k
0inputs+0outputs (0major+1745minor)pagefaults 0swaps
```

elapsed time is just **0.14** sec.

And many command, such as `ping`, `wget`, `curl`, `traceroute`, `pacman`, `apt`, will cost to much time in resolving.

Other tools problem: 
- `netstat`: shows nothing.
- `hexo server` and other programs: Can't bind any port. Or the ports they bind can't be access.

**Comment by rodrymbo** [comment]
It might be good to compare `nslookup` to `dig` and `host`, as well as `/mnt/c/Windows/System32/nslookup.exe` (interop).

It might be good to compare `ping` with `/mnt/c/Windows/System32/PING.EXE` (interop). (On my machine, on one test, the interop took half as much time as the "native" linux ping, but the sample size was too small to be significant.)

It might be good to try `ping` and `/mnt/c/Windows/System32/PING.EXE` on just the IP address(es), to see if name resolution is a factor, or if one IP address pings faster than the other. One might expect `www.bing.com` to be cached for a Windows 10 machine by whatever is doing name resolution, but you never know. One might also expect that the second time you try to ping it, it would be at least a little faster because the second time the lookups should come from a cache. Or if one or more of the DNS resolvers being used are having network problems, the resolution could take a variable amount of time whether from a cache (or not). So a ping to the IP address(es) would eliminate name resolution as a variable.

(As a general rule, if name resolution were coming from a local caching name resolver, one might expect faster response than from a name server somewhere on the Public Internet, but unless you set up a local caching DNS resolver yourself, `/etc/resolv.conf` is likely to point `ping` to use to a resolver out on the 'net somewhere. Plus `ping` with a name attempts name resolution to find the IP address, then reverse resolution to find the name that goes with the IP address, which might be different, and neither of those lookups is guaranteed to be trivial every time.)

For me, `/mnt/c/Windows/System32/NETSTAT.EXE` gives better information than `/bin/netstat`. I haven't fiddled with it to see if `/bin/netstat` is just showing the (lack of) connections within the WSL subsystem, instead of showing all connections to the Windows machine. Again, the fact that it works at all is what is amazing.

The question of localhost:4000 might be a firewall issue, or lots of other things, I suppose. `/mnt/c/Windows/System32/NETSTAT.EXE -an | grep 4000` could show you if anything is listening on port 4000, while ps could (with appropriate options) tell you if the service is even running. 

That's just me being curious. Not really offering a workaround or diagnosis.

**Comment by Ferouden** [comment]
@rodrymbo 
`dig` costs 0.02 sec and `host` costs 0.71 sec.
And in windows, every command perform as good as it should be.
But in WSL on my PC, `ping`,`wget`,`curl`,`pacman`,`apt`...... cost very much time in resolving domain name.(more than 5 sec)

And about port bind. Even if I am downloading something, `netstat` still shows nothing. And I can't bind any port in WSL, though the program is running.

*May be the problem is I am using a newest version Linux(Ubuntu 17.10, Archlinux)? And some packages are not supported by WSL?*

**Comment by rodrymbo** [comment]
If nslookup(.exe) shows a different "Server" in Windows compared to WSL, that could explain the difference in resolution time.  Otherwise, it's either caching or something in the way things in WSL resolve DNS names. It may be that Windows (for example) is caching the lookup after the first time, but WSL individual programs resolve things by querying a DNS server. One would hope the DNS server is caching, and that it is "nearby" enough that network time is negligible, but it's something to check.

My LAN has a caching dns resolver that is used by both Windows and WSL, and I can't tell the difference (since it is local and fast) in how long it takes programs on either side to resolve things -- especially after the first request, which assures that the name is already in the cache for subsequent requests. 

As for Netstat: NETSTAT.EXE will be the gold standard for what is listening or not, since all WSL networking (currently) goes through Windows. If something is open in WSL it will show up in the windows NETSTAT.EXE.

Did you try opening the port in Windows Defender Firewall (or whatever firewall you are using)? You might have to open the port for all programs, since the firewall sometimes has trouble opening for a specific WSL program (though that seems better recently). I can't recall whether localhost is supposed to be open automatically by the firewall or not. There are instructions around for getting sshd running in WSL (after stopping the Windows service if it is already running), and I've heard of lots of people who do get it working without a lot of trouble. Maybe following those instructions will get a port open in WSL that you can see in NETSTAT.EXE.

The other variable to eliminate, in my experience, is ipv6. Sometimes I say "localhost" and the program listens on `::1` instead of `127.0.0.1` as I expect, and I'm troubleshooting ipv4 things and none of them works. 

The WSL environment is mostly agnostic about what Linux binaries it is trying to run (I think). But by all means try switching to one of the versions in the Windows Store for comparison, to see if you are indeed identifying incompatibilities with 17.10 -- I'm sure the devs would want to know about that.

**Comment by DarthSpock** [comment]
@Ferouden Just so you are aware, there is a beta version of 18.04 LTS available, so 17.10 can't be considered the newest Ubuntu image. You are complaining about latency but you're missing the bigger picture of that for the longest time since WSL was born, there were no networking features at all and now they work! Only major hurtle left for networking has to do with AF_Packet for networking monitoring/vulnerability management/pen testing. The performance of WSL isn't as good as native Windows OR Linux and won't be for some time. The WSL team is actively working on improving performance for everything as compiling anything is still slower than native Linux. This *should* have no bearing on what version of Ubuntu you are running as this is more specific to the underlying infrastructure of WSL itself and affects all distros running under WSL. About packages not being supported by WSL, there are many packages that have yet to be supported by WSL that are part of 16.04 Xenial, most of these packages not being supported is due to dependency for X. There are ways to bypass of course and force most things through the terminal but others that you have to run an unsupported, unofficial method of running X on Windows via Xming or something similar to get running. If there are certain packages you feel ought to be supported but aren't (regardless of the version of Linux you are running), create individual tickets for each of those packages and have a specific use case in mind so the devs can understand where you are coming from. If it's requiring a feature not yet implemented, then you will need to create a Uservoice if one doesn't exist that addresses your issue. If there is one that does already, then upvote it.
